File: card-infra/src/main/java/com/jean202/cardmizer/infra/persistence/InMemorySpendingRecordAdapter.java
package com.jean202.cardmizer.infra.persistence;

import com.jean202.cardmizer.core.application.ResourceNotFoundException;
import com.jean202.cardmizer.core.domain.CardId;
import com.jean202.cardmizer.core.domain.SpendingPeriod;
import com.jean202.cardmizer.core.domain.SpendingRecord;
import com.jean202.cardmizer.core.port.out.DeleteSpendingRecordPort;
import com.jean202.cardmizer.core.port.out.LoadSpendingRecordsByCardAndPeriodPort;
import com.jean202.cardmizer.core.port.out.LoadSpendingRecordsPort;
import com.jean202.cardmizer.core.port.out.SaveSpendingRecordPort;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.CopyOnWriteArrayList;

public class InMemorySpendingRecordAdapter implements
        LoadSpendingRecordsPort,
        SaveSpendingRecordPort,
        LoadSpendingRecordsByCardAndPeriodPort,
        DeleteSpendingRecordPort {

    private final CopyOnWriteArrayList<SpendingRecord> spendingRecords = new CopyOnWriteArrayList<>();
    private final Set<UUID> deletedIds = new HashSet<>();

    public InMemorySpendingRecordAdapter() {
    }

    public InMemorySpendingRecordAdapter(List<SpendingRecord> initialSpendingRecords) {
        spendingRecords.addAll(initialSpendingRecords);
    }

    @Override
    public List<SpendingRecord> loadByPeriod(SpendingPeriod period) {
        return spendingRecords.stream()
                .filter(record -> !deletedIds.contains(record.id()))
                .filter(record -> period.includes(record.spentOn()))
                .sorted(Comparator.comparing(SpendingRecord::spentOn).thenComparing(SpendingRecord::id))
                .toList();
    }

    @Override
    public List<SpendingRecord> loadByPeriodAndCard(SpendingPeriod period, CardId cardId) {
        return spendingRecords.stream()
                .filter(record -> !deletedIds.contains(record.id()))
                .filter(record -> period.includes(record.spentOn()))
                .filter(record -> cardId == null || record.cardId().equals(cardId))
                .sorted(Comparator.comparing(SpendingRecord::spentOn).reversed()
                        .thenComparing(Comparator.comparing(SpendingRecord::id).reversed()))
                .toList();
    }

    @Override
    public void save(SpendingRecord spendingRecord) {
        spendingRecords.add(spendingRecord);
    }

    @Override
    public void delete(UUID id) {
        boolean exists = spendingRecords.stream().anyMatch(record -> record.id().equals(id));
        if (!exists || deletedIds.contains(id)) {
            throw new ResourceNotFoundException("Spending record not found: " + id);
        }
        deletedIds.add(id);
    }
}

File: card-infra/src/main/kotlin/com/jean202/cardmizer/infra/persistence/InMemorySpendingRecordAdapter.kt
package com.jean202.cardmizer.infra.persistence

import com.jean202.cardmizer.core.application.ResourceNotFoundException
import com.jean202.cardmizer.core.domain.CardId
import com.jean202.cardmizer.core.domain.SpendingPeriod
import com.jean202.cardmizer.core.domain.SpendingRecord
import com.jean202.cardmizer.core.port.out.DeleteSpendingRecordPort
import com.jean202.cardmizer.core.port.out.LoadSpendingRecordsByCardAndPeriodPort
import com.jean202.cardmizer.core.port.out.LoadSpendingRecordsPort
import com.jean202.cardmizer.core.port.out.SaveSpendingRecordPort
import java.util.UUID
import java.util.concurrent.CopyOnWriteArrayList

class InMemorySpendingRecordAdapter(
    initialSpendingRecords: List<SpendingRecord> = emptyList(),
) : LoadSpendingRecordsPort, SaveSpendingRecordPort, LoadSpendingRecordsByCardAndPeriodPort, DeleteSpendingRecordPort {

    private val spendingRecords = CopyOnWriteArrayList(initialSpendingRecords)
    private val deletedIds = mutableSetOf<UUID>()

    override fun loadByPeriod(period: SpendingPeriod): List<SpendingRecord> =
        spendingRecords
            .filter { it.id !in deletedIds && period.includes(it.spentOn) }
            .sortedWith(compareBy({ it.spentOn }, { it.id }))

    override fun loadByPeriodAndCard(period: SpendingPeriod, cardId: CardId?): List<SpendingRecord> =
        spendingRecords
            .filter { it.id !in deletedIds && period.includes(it.spentOn) && (cardId == null || it.cardId == cardId) }
            .sortedWith(compareByDescending<SpendingRecord> { it.spentOn }.thenByDescending { it.id })

    override fun save(spendingRecord: SpendingRecord) {
        spendingRecords.add(spendingRecord)
    }

    override fun delete(id: UUID) {
        val exists = spendingRecords.any { it.id == id }
        if (!exists || id in deletedIds) {
            throw ResourceNotFoundException("Spending record not found: $id")
        }
        deletedIds.add(id)
    }
}
